### PR TITLE
Commit Utf8JsonWriter contents to buffer before writing directly to b…

### PR DESCRIPTION
**Pull request by ElizabethOkerio**
### Issues

*This pull request fixes #xxx.*

### Description

This PR fixes a bug in the ODataUtf8JsonWriter's `StartStreamValueScopeAsync` where the contents of the utf8jsonwriter are not committed to the buffer before writing stream content to the buffer directly. This leads to content previously written by the utf8jsonwriter being overwritten which leads to corrupt data. Committing utf8jsonwriter's content to buffer before streaming data to the buffer directly fixes this issue. 

The PR also has logic to  add support for float.PositiveInfinity ( -> "INF"), float.NegativityInfinity (-> "-INF") and float.NaN(->"NaN") to ODataUtf8JsonWriter. Support was already there for double, but not float(an exception would be thrown whenUtf8JsonWritertries to writefloat.NaN` for example)

The ODataUtf8JsonWriter.TextWriter did not override Write(char), This PR has changes to override both TextWriter.Write(char) and TextWriter.WriteAsync(char).

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs [hunk 1]
```diff
@@ -60,9 +60,11 @@ public void EndStreamValueScope()
         /// <returns>A task representing the asynchronous operation. The task result contains a stream for writing the stream value.</returns>
         public async Task<Stream> StartStreamValueScopeAsync()
         {
-            this.WriteSeparatorIfNecessary();
+            this.CommitUtf8JsonWriterContentsToBuffer();
+            this.WriteItemWithSeparatorIfNeeded();
             this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
-            await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+            await this.FlushAsync().ConfigureAwait(false);
```
Comment (habbes) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs:67]: Not a request for change, but I wonder why opted for `FlushAsync()` instead of `FlushIfBufferThresholdReachedAsync()` here.

**Review by habbes — APPROVED** [review]

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs (hunk 1 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs:67]: No reason really but I think it'll be good to flush when buffer threshold is reached instead of flushing every other time

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs [hunk 2]
```diff
@@ -198,9 +210,53 @@ public override async ValueTask DisposeAsync()
...
+                if (this.singleCharBuffer != null)
+                {
+                    ArrayPool<char>.Shared.Return(this.singleCharBuffer);
+                }
+
                 await this.FlushAsync().ConfigureAwait(false);
             }
 
+            /// <summary>
+            /// Writes the specifid character to the ODataUtf8JsonWriter.
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs:222]:
```suggestion
            /// Writes the specified character to the ODataUtf8JsonWriter.
```

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs [hunk 3]
```diff
@@ -60,8 +60,10 @@ public void EndStreamValueScope()
         /// <returns>A task representing the asynchronous operation. The task result contains a stream for writing the stream value.</returns>
         public async Task<Stream> StartStreamValueScopeAsync()
         {
-            this.WriteSeparatorIfNecessary();
+            this.CommitUtf8JsonWriterContentsToBuffer();
+            this.WriteItemWithSeparatorIfNeeded();
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs:64]: Why the change from `WriteSeparatorIfNecessary` to `WriteItemWithSeparatorIfNeeded`? Which item would get written in this case?

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs [hunk 4]
```diff
@@ -142,8 +142,15 @@ private bool CheckIfWritingJson(string currentContentType)
         internal sealed class ODataUtf8JsonTextWriter : TextWriter
         {
             private readonly ODataUtf8JsonWriter jsonWriter = null;
+            // Buffer used to store chars that were could not be encoded due to
+            // insufficient data in the input buffer. The chars will be prepended
+            // to the next chunk of input.
             private char[] buffer;
+            // Number of chars in the buffer that are left over frevious chunk.
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs:149]:
```suggestion
            // Number of chars in the buffer that are left over previous chunk.
```
Also are you missing "from" after "over"?

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs [hunk 5]
```diff
@@ -198,9 +210,53 @@ public override async ValueTask DisposeAsync()
...
+                }
+                else
+                {
+                    ReadOnlySpan<char> charToWrite = stackalloc char[1] { value };
+                    this.WriteCharsInChunksWithoutEscaping(charToWrite);
+                }
+            }
+
+            /// <summary>
+            /// Writes the specifid character to the ODataUtf8JsonWriter.
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs:240]:
```suggestion
            /// Asynchronously writes the specified character to the ODataUtf8JsonWriter.
```

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs [hunk 6]
```diff
@@ -142,8 +142,15 @@ private bool CheckIfWritingJson(string currentContentType)
         internal sealed class ODataUtf8JsonTextWriter : TextWriter
         {
             private readonly ODataUtf8JsonWriter jsonWriter = null;
+            // Buffer used to store chars that were could not be encoded due to
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs:145]: Since you decided to describe what the private fields do, it'd be better to use `/// <summary></summary>`. That way the intellisense feature display the description wherever the field is used

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs [hunk 7]
```diff
@@ -142,8 +142,15 @@ private bool CheckIfWritingJson(string currentContentType)
...
         {
             private readonly ODataUtf8JsonWriter jsonWriter = null;
+            // Buffer used to store chars that were could not be encoded due to
+            // insufficient data in the input buffer. The chars will be prepended
+            // to the next chunk of input.
             private char[] buffer;
+            // Number of chars in the buffer that are left over frevious chunk.
             private int numOfCharsNotWrittenFromPreviousChunk = 0;
+            // This buffer is used to by Write(char) to store the char so
+            // that we can re-use our Write(char[], ...) method.
```
Comment (gathogojr) [inline src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs:152]:
```suggestion
            // that we can reuse the methods for writing chars in chunks.
```

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonWriterAsyncTests.cs [hunk 8]
```diff
@@ -383,6 +383,39 @@ private async Task VerifyWriterPrimitiveValueWithIeee754CompatibleAsync<T>(T par
...
         }
 
+        [Theory]
+        [InlineData(124.45, "124.45")]
+        // We write the .0 for doubles without a fractional part
+        // for consistency with the original JsonWriter implementation
+        // and with previous versions. However, it's not a hard requirement.
+        // Clients should not rely on the .0 to decide whether a value
+        // is an integer or double.
+        // In the future we can consider relaxing the need to add a .0 (possibly behind a feature flag).
```
Comment (gathogojr) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonWriterAsyncTests.cs:393]: Maybe we should create an issue to streamline this in ODL 9. The issue would also serve to document our intentions...

Comment (ElizabethOkerio) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/Json/JsonWriterAsyncTests.cs:393]: We didn't make this change in 7.x.We'll make it in 8.x.

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs (hunk 3 above)
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs:64]: It's still the same separator, it's just that `WriteItemWithSeparatorIfNeeded` has logic responsible for writing a separator before an item if the item is within an array and it is not the first item.

File: src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs [hunk 9]
```diff
@@ -198,9 +210,53 @@ public override async ValueTask DisposeAsync()
...
+                }
+            }
+
+            /// <summary>
+            /// Writes the specifid character to the ODataUtf8JsonWriter.
+            /// </summary>
+            /// <param name="value">The character to write.</param>
+            public override async Task WriteAsync(char value)
+            {
+                this.singleCharBuffer ??= ArrayPool<char>.Shared.Rent(1);
```
[the remark this answers is not in the document]
Comment (ElizabethOkerio) [reply src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs:251]: This is a performance optimization thing. `stackalloc` allocates memory on the stack which is fast and automatically freed when method returns.  The synchronous `write` method completes it's work in a single, uninterrupted flow of execution. This is not the case in the asynchronous method which can yield control back to called and continue with execution later. The buffer in this case will need to exist across asynchronous waits which will require something more durable than a stack hence the use of the `ArrayPool`

**Review by gathogojr — APPROVED** [review]

**Review by habbes — APPROVED** [review]
